File: backend/app/models/cafe.py
from sqlalchemy import Column, String, Float, Text
from app.database import Base


class Cafe(Base):
    __tablename__ = "cafes"

    id = Column(String, primary_key=True)
    name = Column(String, nullable=False, index=True)
    city = Column(String, nullable=False, index=True)
    district = Column(String, index=True)  # 行政區 (e.g. "大安區")
    address = Column(String)
    latitude = Column(Float)
    longitude = Column(Float)
    url = Column(String)
    mrt = Column(String)  # Normalized MRT station name
    open_time = Column(Text)

    # Cafe Nomad ratings (0-5 scale, kept for data integrity)
    wifi = Column(Float, default=0)
    socket = Column(Float, default=0)
    quiet = Column(Float, default=0)
    tasty = Column(Float, default=0)
    cheap = Column(Float, default=0)
    music = Column(Float, default=0)
    seat = Column(Float, default=0)

    # Special features
    limited_time = Column(String)  # "yes", "no", "maybe"
    standing_desk = Column(String)  # "yes", "no"
    has_reservation = Column(String, default="")  # "yes", "no", ""

File: frontend/src/types/cafe.ts
export interface Cafe {
  id: string;
  name: string;
  city: string;
  address: string;
  district?: string;
  latitude: number;
  longitude: number;
  url: string;
  mrt: string;
  mrt_station?: string;
  bus_stop?: string | null;
  open_time: string;
  wifi: number;
  socket: number;
  quiet: number;
  tasty: number;
  cheap: number;
  music: number;
  seat: number;
  price?: number | null;
  quiet_level?: "quiet" | "normal" | "loud";
  has_wifi?: boolean;
  has_socket?: boolean;
  reservable?: boolean | null;
  limited_time: string;
  standing_desk: string;
}

export interface CafeRecommendation {
  cafe: Cafe;
  score: number;
  distance_km: number | null;
}

export interface District {
  name: string;
  mrt_stations: string[];
}

export interface Area {
  city: string;
  cafe_count: number;
  city_name: string;
  districts: District[];
  mrt_stations: string[];
}

export interface Filters {
  city: string;
  district: string;
  keyword: string;
  mrt_station: string;
  bus_stop: string;
  wifi: boolean;
  socket: boolean;
  quiet_level: "" | "quiet" | "normal" | "loud";
  max_price: number | null;
  limited_time: string;
  reservable: boolean;
}
